File: 2022/Go/day2.go
package main

import (
	"fmt"
	"strings"
)

func DayTwoPartOne() {
	input := readInput("day2.txt")

	om := make(map[string]string)
	pm := make(map[string]string)

	om["A"] = "Rock"
	om["B"] = "Paper"
	om["C"] = "Scissors"
	pm["X"] = om["A"]
	pm["Y"] = om["B"]
	pm["Z"] = om["C"]

	w := make(map[string]string)

	w["Rock"] = "Scissors"
	w["Paper"] = "Rock"
	w["Scissors"] = "Paper"

	ms := make(map[string]int)

	ms["Rock"] = 1
	ms["Paper"] = 2
	ms["Scissors"] = 3

	var score int
	for _, line := range input {
		split := strings.Split(line, " ")
		o := om[split[0]]
		p := pm[split[1]]

		// We lose
		if w[o] == p {
			score += ms[p]
			continue
		}

		// We win
		if w[p] == o {
			score += ms[p]
			score += 6
			continue
		}

		// Draw
		score += ms[p]
		score += 3
	}

	fmt.Println(score)
}

File: 2022/CSharp/Advent/Day2.cs
namespace Advent;

public static class Day2
{
    private static readonly string[] Input = File.ReadAllLines("Day2.txt");

    private const string Rock = "Rock";
    private const string Paper = "Paper";
    private const string Scissors = "Scissors";

    public static void PartOne()
    {
        var score = 0;
        foreach (var line in Input)
        {
            var split = line.Split(' ');
            var opponentMove = OpponentMoves[split[0]];
            var playerMove = PlayerMoves[split[1]];

            // Draw
            if (opponentMove == playerMove)
            {
                score += 3;
                score += GetMovePoints(playerMove);
                
                continue;
            }

            // We win
            if (Moves[playerMove] == opponentMove)
            {
                score += 6;
                score += GetMovePoints(playerMove);

                continue;
            }
            
            // We lose
            score += GetMovePoints(playerMove);
        }

        Console.WriteLine(score);
    }

    public static void PartTwo()
    {
        var score = 0;
        foreach (var line in Input)
        {
            var split = line.Split(' ');
            var opponentMove = OpponentMoves[split[0]];
            string playerMove;
            var outcome = split[1];

            // We should lose
            if (outcome == "X")
            {
                playerMove = Moves.FirstOrDefault(m => m.Key == opponentMove).Value;
                score += GetMovePoints(playerMove);

                continue;
            }

            // We should win
            if (outcome == "Z")
            {
                playerMove = Moves.FirstOrDefault(m => m.Value == opponentMove).Key;
                score += GetMovePoints(playerMove);
                score += 6;

                continue;
            }

            // We should tie
            score += GetMovePoints(opponentMove);
            score += 3;
        }

        Console.WriteLine(score);
    }

    private static int GetMovePoints(string move)
    {
        return move switch
        {
            Rock => 1,
            Paper => 2,
            Scissors => 3,
            _ => throw new ArgumentOutOfRangeException(nameof(move), move, null)
        };
    }

    public static Dictionary<string, string> Moves = new()
    {
        { Rock, Scissors },
        { Paper, Rock },
        { Scissors, Paper }
    };

    public static Dictionary<string, string> OpponentMoves = new()
    {
        { "A", Rock },
        { "B", Paper },
        { "C", Scissors }
    };

    public static Dictionary<string, string> PlayerMoves = new()
    {
        { "X", Rock },
        { "Y", Paper },
        { "Z", Scissors }
    };
}